Index: u-boot-lx2162a-bsp0.4-lx2162au26z-minimal/drivers/usb/host/xhci-ring.c
===================================================================
--- u-boot-lx2162a-bsp0.4-lx2162au26z-minimal.orig/drivers/usb/host/xhci-ring.c
+++ u-boot-lx2162a-bsp0.4-lx2162au26z-minimal/drivers/usb/host/xhci-ring.c
@@ -575,10 +575,13 @@ int xhci_bulk_tx(struct usb_device *udev
 	int ret;
 	u32 trb_fields[4];
 	u64 val_64 = (uintptr_t)buffer;
+	void *last_transfer_trb_addr;
+	int available_length;
 
 	debug("dev=%p, pipe=%lx, buffer=%p, length=%d\n",
 		udev, pipe, buffer, length);
 
+	available_length = length;
 	ep_index = usb_pipe_ep_index(pipe);
 	virt_dev = ctrl->devs[slot_id];
 
@@ -698,7 +701,7 @@ int xhci_bulk_tx(struct usb_device *udev
 		trb_fields[2] = length_field;
 		trb_fields[3] = field | (TRB_NORMAL << TRB_TYPE_SHIFT);
 
-		queue_trb(ctrl, ring, (num_trbs > 1), trb_fields);
+		last_transfer_trb_addr = queue_trb(ctrl, ring, (num_trbs > 1), trb_fields);
 
 		--num_trbs;
 
@@ -711,6 +714,7 @@ int xhci_bulk_tx(struct usb_device *udev
 
 	giveback_first_trb(udev, ep_index, start_cycle, start_trb);
 
+again:
 	event = xhci_wait_for_event(ctrl, TRB_TRANSFER);
 	if (!event) {
 		debug("XHCI bulk transfer timed out, aborting...\n");
@@ -719,6 +723,15 @@ int xhci_bulk_tx(struct usb_device *udev
 		udev->act_len = 0;
 		return -ETIMEDOUT;
 	}
+
+	if ((uintptr_t)(le64_to_cpu(event->trans_event.buffer))	!=
+			(uintptr_t)virt_to_phys(last_transfer_trb_addr)) {
+		available_length -=
+			(int)EVENT_TRB_LEN(le32_to_cpu(event->trans_event.transfer_len));
+		xhci_acknowledge_event(ctrl);
+		goto again;
+	}
+
 	field = le32_to_cpu(event->trans_event.flags);
 
 	BUG_ON(TRB_TO_SLOT_ID(field) != slot_id);
@@ -726,7 +739,7 @@ int xhci_bulk_tx(struct usb_device *udev
 	BUG_ON(*(void **)(uintptr_t)le64_to_cpu(event->trans_event.buffer) -
 		buffer > (size_t)length);
 
-	record_transfer_result(udev, event, length);
+	record_transfer_result(udev, event, available_length);
 	xhci_acknowledge_event(ctrl);
 	xhci_inval_cache((uintptr_t)buffer, length);
 
